xsplice: Only lookup new symbols from xsplice symtabs

If a function is patched multiple times, it will appear in several
symbol tables. However, when resolving symbols, it must always link
against the object which first introduced the symbol.

diff --git a/xen/common/xsplice.c b/xen/common/xsplice.c
index 1fe8fde..e7aea4a 100644
--- a/xen/common/xsplice.c
+++ b/xen/common/xsplice.c
@@ -901,6 +901,7 @@ static int build_symbol_table(struct payload *payload, struct xsplice_elf *elf)
             symtab[nsyms].name = strtab + strtab_len;
             symtab[nsyms].size = elf->sym[i].sym->st_size;
             symtab[nsyms].value = elf->sym[i].sym->st_value;
+            symtab[nsyms].flags = 0;
             strtab_len += strlcpy(strtab + strtab_len, elf->sym[i].name,
                                   KSYM_NAME_LEN) + 1;
             nsyms++;
@@ -930,6 +931,7 @@ static int build_symbol_table(struct payload *payload, struct xsplice_elf *elf)
                 xfree(strtab);
                 return -EEXIST;
             }
+            symtab[i].flags |= XSPLICE_SYMBOL_NEW;
             printk(XENLOG_DEBUG "%s: new symbol %s\n", elf->name,
                    symtab[i].name);
         }
@@ -980,11 +982,11 @@ static int load_payload_data(struct payload *payload, uint8_t *raw, ssize_t len)
     if ( rc )
         goto err_payload;
 
-    rc = build_symbol_table(payload, &elf);
+    rc = find_special_sections(payload, &elf);
     if ( rc )
         goto err_payload;
 
-    rc = find_special_sections(payload, &elf);
+    rc = build_symbol_table(payload, &elf);
     if ( rc )
         goto err_payload;
 
@@ -1409,6 +1411,9 @@ uint64_t xsplice_symbols_lookup_by_name(const char *symname)
     {
         for ( i = 0; i < data->nsyms; i++ )
         {
+            if ( !(data->symtab[i].flags & XSPLICE_SYMBOL_NEW) )
+                continue;
+
             if ( !strcmp(data->symtab[i].name, symname) )
             {
                 value = data->symtab[i].value;
diff --git a/xen/include/xen/xsplice.h b/xen/include/xen/xsplice.h
index bf19f4e..1063244 100644
--- a/xen/include/xen/xsplice.h
+++ b/xen/include/xen/xsplice.h
@@ -28,10 +28,13 @@ struct xsplice_build_id {
    unsigned int len;
 };
 
+#define XSPLICE_SYMBOL_NEW    0x1
+
 struct xsplice_symbol {
     const char *name;
     uint64_t value;
     size_t size;
+    int flags;
 };
 
 int xsplice_control(struct xen_sysctl_xsplice_op *);
